pkg/controller/constraint/constraint_controller.go

```diff
@@ -567,6 +477,135 @@ func (r *ReconcileConstraint) reportErrorOnConstraintStatus(ctx context.Context,
 	return err
 }
 
+func (r *ReconcileConstraint) generateVAPB(ctx context.Context, enforcementAction util.EnforcementAction, instance *unstructured.Unstructured, status *constraintstatusv1beta1.ConstraintPodStatus) (time.Duration, error) {
+	ret := time.Duration(0)
+	if !operations.IsAssigned(operations.Generate) {
+		log.Info("generate operation is not assigned, ValidatingAdmissionPolicyBinding resource will not be generated")
+		return ret, nil
+	}
+	ct := &v1beta1.ConstraintTemplate{}
+	err := r.reader.Get(ctx, types.NamespacedName{Name: strings.ToLower(instance.GetKind())}, ct)
+	if err != nil {
+		return ret, err
+	}
+
+	generateVAPB, VAPEnforcementActions, err := shouldGenerateVAPB(*DefaultGenerateVAPB, enforcementAction, instance)
+	if err != nil {
+		log.Error(err, "could not determine if ValidatingAdmissionPolicyBinding should be generated")
+		return ret, r.reportErrorOnConstraintStatus(ctx, status, err, "could not determine if ValidatingAdmissionPolicyBinding should be generated")
+	}
+	isAPIEnabled := false
+	var groupVersion *schema.GroupVersion
+	if generateVAPB {
+		isAPIEnabled, groupVersion = transform.IsVapAPIEnabled(&log)
+	}
+	if generateVAPB {
+		if !isAPIEnabled {
+			log.Error(ErrValidatingAdmissionPolicyAPIDisabled, "Cannot generate ValidatingAdmissionPolicyBinding", "constraint", instance.GetName())
+			_ = r.reportErrorOnConstraintStatus(ctx, status, ErrValidatingAdmissionPolicyAPIDisabled, "cannot generate ValidatingAdmissionPolicyBinding")
+			generateVAPB = false
+		} else {
+			unversionedCT := &templates.ConstraintTemplate{}
+			if err := r.scheme.Convert(ct, unversionedCT, nil); err != nil {
+				return ret, r.reportErrorOnConstraintStatus(ctx, status, err, "could not convert ConstraintTemplate to unversioned")
+			}
+			hasVAP, err := ShouldGenerateVAP(unversionedCT)
+			switch {
+			case errors.Is(err, celSchema.ErrCodeNotDefined):
+				generateVAPB = false
```

Comment:
Should this error be swallowed? Or should we tell the user that the constraint template does not support VAP?

Reply:
This will mean when we switch the default to always generate VAP and VAPB, we will be throwing error for existing users using constraint template with Rego. I can also see the other use-case, where we would want to notify users if the intent is to use VAP but there is no CEL in the CT.

@ritazh @sozercan Any thoughts on this?

Reply:
Did we not switch to a model where users __must__ use the VAP enforcement point to use VAP? I remember having that discussion. Part of the reason for requiring explicit user intent was to avoid questions like this.

Reply:
In any case, I think an error or "warning"-type error makes sense. Users can use scoped EA to disable it.

Contrast that with users getting no notice that a Rego-sourced CT is incompatible with VAP.

Reply:
IMO we should check if engine is K8sNativeValidation, then log and report the error. I recall this error was swallowed in 3.17.1 because 3.17.0 was reporting this error for existing rego CTs.

Reply:
I do not recall having a discussion where user must use the VAP enforcement point to use VAP. I thought eventually we wanted to enable VAP enforcement point by default in-line with other enforcement points.

Reporting error only when VAP enforcement point explicitly mentioned in constraint through `scopedEnforcementAction` and template does not satisfy conditions to create VAP could work. This will not affect current rego users. And when we switch to VAP by default, the guidance for constraint authors would be:
- If VAP is available to use, GK will use VAP with all constraint. Otherwise GK will use webhook for enforcement. No errors reported on CT about not generating VAP.
- If VAP is not available to use and user wants to use VAP with `scopedEnforcementPoints`, GK reports error on CT and enforces through webhook if included as enforcement point.

Does this make sense?^^

Reply:
As per the community discussion, I will follow up with instroducing field on constraint status for enforcementPoint status under bypod.